### PR TITLE
WPT: Add remaining url tests

### PR DESCRIPTION
**Pull request by npaun**
Adds the following features to the test harness:
- A single `run` entry point
- An `expectedFailures` option to document which tests we know aren't working yet
- A `skippedTests` option for tests that would make workerd crash
- A `verbose` option to print subtests before they are run.

Implements more WPT assertion/test functions:
- subsetTestByKey
- promise_test (probably still needs further improvements)
- assert_not_equals
- assert_true / assert_falwse
- assert_array_equals
- assert_throws_js

Integrates the following WPT tests:
- [X] url-constructor.any.js
    - [ ] Failure: 'Parsing: <http://example.com/\uD800\uD801\uDFFE\uDFFF\uFDD0\uFDCF\uFDEF\uFDF0\uFFFE\uFFFF?\uD800\uD801\uDFFE\uDFFF\uFDD0\uFDCF\uFDEF\uFDF0\uFFFE\uFFFF> without base'
- [X] url-origin.any.js
- [X] url-searchparams.any.js
- [X] url-setters-stripping.any.js
- [X] url-setters.any.js
- [X] url-statics-parse.any.js
- [X] url-tojson.any.js
- [X] urlencoded-parser.any.js
   - [ ] Failure: Request tests fail because WPT is using a bogus method named "LADIDA" and we throw
   - [ ] Failure: Response tests fail because we  need an IO context
- [X] urlsearchparams-append.any.js
   - [ ] Failure: 'URLSearchParams constructor, DOMException as argument'
   - [ ] Failure: 'Construct with 2 unpaired surrogates (no trailing)'
   - [ ] Failure: 'Construct with 3 unpaired surrogates (no leading)'
   - [ ] Failure: 'Construct with object with NULL, non-ASCII, and surrogate keys'
- [X] urlsearchparams-delete.any.js
- [X] urlsearchparams-foreach.any.js
    - [ ] Segfault:  'For-of Check'
- [X] urlsearchparams-get.any.js
- [X] urlsearchparams-getall.any.js
- [X] urlsearchparams-has.any.js
- [X] urlsearchparams-set.any.js
- [X] urlsearchparams-size.any.js
- [X] urlsearchparams-sort.any.js
   - Basically the issue here is sorting between a ligature and an emoji
   - [ ] Failure: Parse and sort: ﬃ&🌈
   - [ ] Failure: URL parse and sort: ﬃ&🌈
- [X] urlsearchparams-stringifier.any.js

### CODE REVIEW

**Comment by npaun** [comment]
@jasnell, @anonrig: I marked this PR as ready for review, because it's now passing CI once the expectedFailures, and skippedTests are set.

I tried to make some brief notes as to what was failing on each test.

File: build/wpt_test.bzl [hunk 1]
```diff
@@ -89,10 +89,11 @@ def generate_external_modules(files):
 
     for file in files.to_list():
         file_path = wd_relative_path(file)
-        if file.basename.endswith(".js"):
+        if file.extension == "js":
             entry = """(name = "{}", esModule = embed "{}")""".format(file.basename, file_path)
-        elif file.basename.endswith(".json"):
-            entry = """(name = "{}", json = embed "{}")""".format(file.basename, file_path)
+        elif file.extension == "json":
+            # TODO(soon): This is ugly but it's hard to manipulate paths in bazel
```
Comment (anonrig) [inline build/wpt_test.bzl:95]: Can you make sure TODO comments include actionable items? It's hard to understand what this comment does.

File: src/workerd/api/wpt/url-test.js [hunk 2]
```diff
@@ -2,20 +2,138 @@
...
-    await import('url-origin.any.js');
-    harness.validate();
-  },
-};
+import { run } from 'harness';
+
+// TODO(soon): Do we care about anything in here?
+//export const historical = run('historical.any.js');
+
+// TODO(soon): we return a different string than theirs
```
Comment (anonrig) [inline src/workerd/api/wpt/url-test.js:10]:
```suggestion
```

File: src/workerd/api/wpt/url-test.js [hunk 3]
```diff
@@ -2,20 +2,138 @@
...
+export const urlOrigin = run('url-origin.any.js');
+export const urlSearchParams = run('url-searchparams.any.js');
+export const urlSettersStripping = run('url-setters-stripping.any.js');
+export const urlSetters = run('url-setters.any.js');
+export const urlStaticsCanParse = run('url-statics-canparse.any.js');
+export const urlStaticsParse = run('url-statics-parse.any.js');
+export const urlToJson = run('url-tojson.any.js');
+
+// TODO(soon): request tests: These tests use an invalid method named "LADIDA", but the tests assume this won't cause any problems
+// TODO(soon): response tests: seem to need an IO context
```
Comment (anonrig) [inline src/workerd/api/wpt/url-test.js:26]:
```suggestion
```

File: src/workerd/api/wpt/url-test.js [hunk 4]
```diff
@@ -2,20 +2,138 @@
...
+  ],
+});
+
+export const urlSearchParamsAppend = run('urlsearchparams-append.any.js');
+
+export const urlSearchParamsConstructor = run(
+  'urlsearchparams-constructor.any.js',
+  {
+    expectedFailures: [
+      // TODO(soon): It's supposed to throw when we call new URLSearchParams(DOMException)
```
Comment (anonrig) [inline src/workerd/api/wpt/url-test.js:108]: The test names are explanatory enough. We don't need such comments unless we deliberately disable the rule due to a limitation of workerd.

File: src/workerd/api/wpt/url-test.js [hunk 5]
```diff
@@ -2,20 +2,130 @@
...
-  async test() {
-    harness.prepare();
-    await import('url-origin.any.js');
-    harness.validate();
-  },
-};
+import { run } from 'harness';
+
+// TODO(soon): Do we care about anything in here?
+//export const historical = run('historical.any.js');
```
Comment (anonrig) [inline src/workerd/api/wpt/url-test.js:8]: This needs to be enabled.

**Review by anonrig — APPROVED** [review]
